modules/bm_sim/src/lookup_structures.cpp

```diff
@@ -0,0 +1,247 @@
+
+
+namespace bm {
+namespace {  // anonymous
+
+static_assert(sizeof(value_t) == sizeof(internal_handle_t),
+              "Invalid type sizes");
+
+// We don't need or want to export these classes outside of this
+// compilation unit.
+
+class LPMTrie : public LookupStructure<LPMMatchKey> {
+ public:
+  explicit LPMTrie(size_t key_width_bytes)
+    : key_width_bytes(key_width_bytes) {
+      trie = bf_lpm_trie_create(key_width_bytes, true);
+    }
+
+  /* Copy constructor */
+  LPMTrie(const LPMTrie& other) = delete;
+
+  /* Move constructor */
+  LPMTrie(LPMTrie&& other) noexcept
+    : key_width_bytes(other.key_width_bytes), trie(other.trie) {}
+
+  ~LPMTrie() {
+    bf_lpm_trie_destroy(trie);
+  }
+
+  /* Copy assignment operator */
+  LPMTrie &operator=(const LPMTrie &other) = delete;
+
+  /* Move assignment operator */
+  LPMTrie &operator=(LPMTrie &&other) noexcept {
+    key_width_bytes = other.key_width_bytes;
+    std::swap(trie, other.trie);
+    return *this;
+  }
+
+  bool lookup(const ByteContainer & key_data,
+              internal_handle_t * handle) override {
+    return bf_lpm_trie_lookup(trie, key_data.data(),
+                              reinterpret_cast<value_t *>(handle));
+  }
+
+  bool entry_exists(const LPMMatchKey & key) override {
+    return bf_lpm_trie_has_prefix(trie, key.data.data(), key.prefix_length);
+  }
+
+  void store_entry(const LPMMatchKey & key,
+                   internal_handle_t handle) override {
+    bf_lpm_trie_insert(trie, key.data.data(), key.prefix_length,
+        (value_t) handle);
+  }
+
+  void delete_entry(const LPMMatchKey & key) override {
+    bf_lpm_trie_delete(trie, key.data.data(), key.prefix_length);
+  }
+
+  void clear() override {
+    bf_lpm_trie_destroy(trie);
+    trie = bf_lpm_trie_create(key_width_bytes, true);
+  }
+
+ private:
+  size_t key_width_bytes{0};
+  bf_lpm_trie_t *trie{nullptr};
+};
+
+class ExactMap : public LookupStructure<ExactMatchKey> {
+ public:
+  explicit ExactMap(size_t size) {
+    entries_map.reserve(size);
+  }
+
+  bool lookup(const ByteContainer & key,
+              internal_handle_t * handle) override {
+    const auto entry_it = entries_map.find(key);
+    if (entry_it == entries_map.end()) {
+      return false;  // Nothing found
+    } else {
+      *handle = entry_it->second;
+      return true;
+    }
+  }
+
+  bool entry_exists(const ExactMatchKey & key) override {
+    (void) key;
+    return entries_map.find(key.data) != entries_map.end();
+  }
+
+  void store_entry(const ExactMatchKey & key,
+      internal_handle_t handle) override {
+    entries_map[key.data] = handle;  // key is copied, which is not great
+  }
+
+  void delete_entry(const ExactMatchKey & key) override {
+    entries_map.erase(key.data);
+  }
+
+  void clear() override {
+    entries_map.clear();
+  }
+
+ private:
+  // TODO(gordon) Should this be internal_handle_t ?
+  std::unordered_map<ByteContainer, entry_handle_t, ByteContainerKeyHash>
```

Comment:
yes I believed this should be `internal_handle_t` and that there was a typo in my original code